### ISSUE TITLE
fullBaseUrl with HTTP_X_FORWARDED_PROTO support

### ISSUE THREAD

**Issue by jicho** [cakephp/cakephp#16393]
### Description

Hi,

When you leave "`'App.fullBaseUrl' => false`" in the app.php and are running CakePHP behind a proxy server that is managing the SSL part of the world you end up with http redirects since CakePHP does not "see" that the site is served over https.

Wouldn't it be an idea to change Router::fullBaseUrl() a little so it "see"/"looks" for the HTTP_X_FORWARDED_PROTO header and let is return a https url instead of http.

I've encountered this, in my case, problem when I was setting up the form-action value in the CSP-headers.
On my production server all the login 302-redirects where blocked by the browser because CakePHP was returning the 302-redirect als http and not https.

CSP-header setting (https://book.cakephp.org/4/en/security/content-security-policy.html):
```
'form-action' => [
     'self' => true,
],
```

The problem in steps:
```
Request:
user login request over httpS -> request send by browser -> proxy server -> (apache) -> CakePHP

Response:
CakePHP -> (apache) -> proxy server -> http-redirect -> browser -> user stays on the login page
```

Funny detail... the login is handled correctly, to when the user goes to a page after logging manually he can see it.

For now I've solved the problem by adding this to the app.php"
```
function getFullBaseUrl()
{
    if(!env('HTTP_HOST')) {
        return false;
    }

    $host = 'http://' . $_SERVER['HTTP_HOST'];

    if(env('https') || (env('HTTP_X_FORWARDED_PROTO') && env('HTTP_X_FORWARDED_PROTO') == 'https')) {
        $host = 'https://' . $_SERVER['HTTP_HOST'];
    }

    return $host;
}
```
In combination with `'fullBaseUrl' => getFullBaseUrl()`.

**Comment by ndm2** [comment]
I'd rather not see more client controllable headers being trusted out of the box.

Adding logic to the bootstrap that is disabled by default, and requires people to explicitly enable it and makes them aware that this should only be done when behind a trusted proxy, could be a compromise.

**Comment by jicho** [comment]
@ndm2 After find the following code in bootstrap.php:

```
/*
 * Set the full base URL.
 * This URL is used as the base of all absolute links.
 */
$fullBaseUrl = Configure::read('App.fullBaseUrl');
if (!$fullBaseUrl) {
    $s = null;
    if (env('HTTPS')) {
        $s = 's';
    }

    $httpHost = env('HTTP_HOST');
    if (isset($httpHost)) {
        $fullBaseUrl = 'http' . $s . '://' . $httpHost;
    }
    unset($httpHost, $s);
}
if ($fullBaseUrl) {
    Router::fullBaseUrl($fullBaseUrl);
}
unset($fullBaseUrl);
```

I agree that this is the place for the change.
For some reason I didn't see this piece of code during my search for a solution.

The proxy addition would be nice to have out-of-the-box when starting a new project :)

From this:
```
    if (env('HTTPS')) {
        $s = 's';
    }
```

To this:
```
    if (env('HTTPS') || (env('HTTP_X_FORWARDED_PROTO') && env('HTTP_X_FORWARDED_PROTO') == 'https')) {
        $s = 's';
    }
```

Does the trick.

**Comment by markstory** [comment]
Closing as the app template has been updated with an opt-in path to trust proxy headers. We can't always trust request headers as not all applications run behind a proxy and using proxy headers in that scenario is not safe.
